### PR TITLE
feat(babel): Auto-inject sentry-label from static text children

### LINKED ISSUE #6098: Auto-inject `sentry-label` from static JSX text children in Babel plugin
Extend `@sentry/babel-plugin-component-annotate` to auto-emit a `sentry-label` prop from static JSX text children at build time, so devs get meaningful labels without writing the prop by hand.

Example transform:

```jsx
// Before
<TouchableOpacity onPress={save}>
  <Text>Save workout</Text>
</TouchableOpacity>

// After (at build time)
<TouchableOpacity sentry-label="Save workout" onPress={save}>
  <Text>Save workout</Text>
</TouchableOpacity>
```

The SDK already reads `sentry-label` via `getLabelValue` in `touchevents.tsx:325` — no SDK-side change needed.

**Constraints:**

* Only static string literals — `<Text>{dynamicVar}</Text>` gets nothing
* Skip if `sentry-label` is already present (user-authored wins)
* Cap at \~64 chars
* Same plugin is used by web JS SDK — confirm no unintended side effects there

**Scope:** Change is in `getsentry/sentry-javascript-bundler-plugins` repo, not this one. Babel AST visitor addition.

### PR DESCRIPTION
**Pull request by antonis**
## Summary

Adds an opt-in `autoInjectSentryLabel` option to `@sentry/babel-plugin-component-annotate` that automatically extracts static text from JSX children and injects a `sentry-label` attribute on the root element.

This enables React Native apps to get meaningful touch breadcrumb labels (e.g., "Save workout", "Add to cart") without manual annotation. The feature walks the JSX tree up to 3 levels deep, collecting text from recognized text components (default: `Text`, `text`), and bails out entirely if any dynamic content is found anywhere in the subtree.

Resolves https://github.com/getsentry/sentry-react-native/issues/6098

### Key design decisions

- **Opt-in only** — `autoInjectSentryLabel` defaults to `false`, no behavior change for existing users. When disabled, the only overhead is two property reads during context creation and one boolean check per `processJSX` call — none of the new extraction/injection code is reached.
- **Conservative bail-out** — any dynamic expression (`{variable}`, `{t('key')}`, template literals) anywhere in the subtree causes the entire label to be skipped — including dynamic content nested inside non-text wrapper elements within text components (e.g., `<Text>Hello <Bold>{name}</Bold></Text>`). This is stricter than the runtime extraction in sentry-react-native#6106, which skips dynamic nodes and continues. Build-time can't evaluate dynamic content, so skipping entirely avoids partial/misleading labels.
- **Hyphenated `sentry-label`** — React strips unknown hyphenated props from DOM elements, so this is safe for web even though the feature targets React Native
- **Cross-SDK safe** — `bundler-plugin-core`'s `createComponentNameAnnotateHooks` does not pass this option through, so only React Native (which configures the Babel plugin directly via Metro) can activate it. Web SDKs are unaffected.
- **Nested `<Text>` support** — handles the standard RN inline styling pattern (`<Text>Hello <Text style={bold}>world</Text></Text>`), recursively extracting text from nested text components
- **Fragment handling** — when the root is a fragment, the first `JSXElement` child is used as the injection target. Fragments are treated as transparent wrappers and do not consume depth budget during text search.
- **Truncation** — labels exceeding 64 characters are truncated with `...`
- **`textComponentNames` option** — configurable list of component names whose children are treated as text content

### New plugin options

| Option | Type | Default | Description |
|--------|------|---------|-------------|
| `autoInjectSentryLabel` | `boolean` | `false` | Enable auto-injection of `sentry-label` |
| `textComponentNames` | `string[]` | `["Text", "text"]` | Component names whose JSXText children are considered text content |

## Test plan

- [x] 55 new tests in `sentry-label.test.ts` covering: opt-in behavior, basic extraction, multiple text children, skip conditions, truncation, depth limit, text component names, nested text (RN inline styling), fragment handling (including depth budget), dynamic content bail-out in non-text wrappers, web compatibility, edge cases, multiple components per file, ternary returns
- [x] All 175 tests pass (55 new + 120 existing unchanged)
- [x] Verified no regressions in existing `test-plugin.test.ts` (65 tests) and `experimental.test.ts` (55 tests)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/babel-plugin-component-annotate/src/index.ts [hunk 1]
```diff
@@ -658,4 +676,198 @@ function getJSXMemberExpressionObjectName(
...
+        }
+        texts.push(...innerTexts);
+      }
+    } else if (t.isJSXSpreadChild(child)) {
+      return null;
+    }
+  }
+
+  return texts;
+}
```
[the remark this answers is not in the document]
Comment (antonis) [reply packages/babel-plugin-component-annotate/src/index.ts:806]: Fixed with ebee4f47ca6517ec3f9f32a1c6d95d0221137d7b

File: packages/babel-plugin-component-annotate/src/index.ts [hunk 2]
```diff
@@ -658,4 +676,204 @@ function getJSXMemberExpressionObjectName(
...
+          depth - 1,
+          false
+        );
+        if (result === null) {
+          return null;
+        }
+        texts.push(...result);
+      }
+    } else if (t.isJSXFragment(child)) {
+      const result = extractStaticTextFromChildren(t, child, textComponentNames, depth - 1, false);
```
[the remark this answers is not in the document]
Comment (antonis) [reply packages/babel-plugin-component-annotate/src/index.ts]: df91d27035bec39d852aa595b82a5aef9c545204

File: packages/babel-plugin-component-annotate/src/index.ts [hunk 3]
```diff
@@ -658,4 +676,204 @@ function getJSXMemberExpressionObjectName(
...
+      }
+    } else if (t.isJSXElement(child)) {
+      const childName = getElementName(t, child.openingElement);
+      if (textComponentNames.includes(childName)) {
+        const innerTexts = extractTextFromTextComponent(t, child, textComponentNames);
+        if (innerTexts === null) {
+          return null;
+        }
+        texts.push(...innerTexts);
+      }
```
[the remark this answers is not in the document]
Comment (antonis) [reply packages/babel-plugin-component-annotate/src/index.ts:793]: df91d27035bec39d852aa595b82a5aef9c545204

File: packages/babel-plugin-component-annotate/src/index.ts [hunk 4]
```diff
@@ -658,4 +676,209 @@ function getJSXMemberExpressionObjectName(
...
+          depth - 1,
+          false
+        );
+        if (result === null) {
+          return null;
+        }
+        texts.push(...result);
+      }
+    } else if (t.isJSXFragment(child)) {
+      const result = extractStaticTextFromChildren(t, child, textComponentNames, depth - 1, isRoot);
```
[the remark this answers is not in the document]
Comment (antonis) [reply packages/babel-plugin-component-annotate/src/index.ts]: Fixed in 0bd679b — fragments now pass depth instead of depth - 1 when recursing, so they don't consume a nesting level.

[the remark this answers is not in the document]
Comment (antonis) [reply packages/babel-plugin-component-annotate/src/index.ts]: Fixed in 0bd679b — fragments no longer decrement the depth counter. Added a test verifying that a fragment at depth 3 doesn't prevent text extraction.

**Comment by timfish** [comment]
The regular component injection plugin breaks props in some circumstances which has stopped us from enabling this plugin by default:
- https://github.com/getsentry/sentry-javascript-bundler-plugins/issues/492

We've recently added an experimental mode which injects the annotations into the HTML elements inside the component:
https://github.com/getsentry/sentry-javascript-bundler-plugins/blob/main/packages/babel-plugin-component-annotate/src/experimental.ts

**Comment by antonis** [comment]
Thank you for having a look and the context @timfish 🙇 

The `sentry-label` attribute is indeed injected on the root JSX element, the same injection point as `data-sentry-component` described in #492. I think this should still work in this case since this option is only reachable from React Native (via direct Babel plugin config in Metro). The `bundler-plugin-core` doesn't pass it through, so web SDKs should not be affected. It is also opt-in and `autoInjectSentryLabel` defaults to `false`.
> We've recently added an experimental mode which injects the annotations into the HTML elements inside the component

I think this is not applicable for RN since there are no HTML elements.

**Review by timfish — COMMENTED** [review]
Rather than separate `autoInjectSentryLabel` and `textComponentNames` options, we should probably combine these so it's more obvious they are always used together and are unrelated to the other options.

Something like this:
```ts
autoInjectSentryLabel?: boolean | { textComponentNames: string[] }
```
`autoInjectSentryLabel` defaults to false. `autoInjectSentryLabel: true` gets transformed to `autoInjectSentryLabel: { textComponentNames: ["Text", "text"] }`?

> since this option is only reachable from React Native

If this option should not be visible to other users of the plugin, we might want to use jsdoc `@hidden` on the options so that users can't discover it.

**Comment by antonis** [comment]
Thank you for you feedback and suggestions @timfish 🙇 

> Rather than separate `autoInjectSentryLabel` and `textComponentNames` options, we should probably combine these so it's more obvious they are always used together and are unrelated to the other options.

Good idea 👍 Combined with 985b9678976a5a6f8ba9e949ceb74577068d1581

> If this option should not be visible to other users of the plugin, we might want to use jsdoc `@hidden` on the options so that users can't discover it.

I've added it on the internal AnnotationOpts interface as a signal in 985b9678976a5a6f8ba9e949ceb74577068d1581

File: packages/babel-plugin-component-annotate/src/index.ts [hunk 5]
```diff
@@ -658,4 +683,209 @@ function getJSXMemberExpressionObjectName(
...
+        if (innerTexts === null) {
+          return null;
+        }
+        texts.push(...innerTexts);
+      } else {
+        const innerTexts = extractTextFromTextComponent(t, child, textComponentNames);
+        if (innerTexts === null) {
+          return null;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (antonis) [reply packages/babel-plugin-component-annotate/src/index.ts:793]: Updated with a46f87889593848b8992ac4d61b942b6cee19ba1

File: packages/babel-plugin-component-annotate/src/index.ts [hunk 6]
```diff
@@ -170,6 +185,11 @@ function createJSXProcessingContext(
     attributeNames: attributeNamesFromState(state),
     ignoredComponents: state.opts.ignoredComponents ?? [],
     fragmentContext: state.sentryFragmentContext,
+    autoInjectSentryLabel: !!state.opts.autoInjectSentryLabel,
+    textComponentNames:
+      (typeof state.opts.autoInjectSentryLabel === "object"
+        ? state.opts.autoInjectSentryLabel.textComponentNames
+        : undefined) ?? DEFAULT_TEXT_COMPONENT_NAMES,
```
[the remark this answers is not in the document]
Comment (antonis) [reply packages/babel-plugin-component-annotate/src/index.ts:192]: Updated with d0cacb6faffa5a37a50b9d59d7b56f6ab9b7f726

File: packages/babel-plugin-component-annotate/src/index.ts [hunk 7]
```diff
@@ -658,4 +683,210 @@ function getJSXMemberExpressionObjectName(
...
+          return null;
+        }
+        texts.push(...innerTexts);
+      } else {
+        const innerTexts = extractTextFromTextComponent(t, child, textComponentNames);
+        if (innerTexts === null) {
+          return null;
+        }
+        texts.push(...innerTexts);
+      }
```
[the remark this answers is not in the document]
Comment (antonis) [reply packages/babel-plugin-component-annotate/src/index.ts:793]: Updated with d0cacb6faffa5a37a50b9d59d7b56f6ab9b7f726

**Comment by timfish** [comment]
I'm struggling to work out how this has broken the component annotation tests on Windows 🤔

**Comment by antonis** [comment]
> I'm struggling to work out how this has broken the component annotation tests on Windows 🤔

I've rebased on main in case this was some caching issue but still fails :( 👀

**Comment by antonis** [comment]
> > I'm struggling to work out how this has broken the component annotation tests on Windows 🤔
> 
> I've rebased on main in case this was some caching issue but still fails :( 👀

Deleting the actual ci caches [worked](https://github.com/getsentry/sentry-javascript-bundler-plugins/actions/runs/25666500315/job/75341448554?pr=925) 🎉 
It seems that the Windows node_modules cache was populated before the Rolldown v1 merge, and subsequent runs kept restoring stale dependencies.

**Review by timfish — APPROVED** [review]
